### PR TITLE
Add updatepeer RPC

### PR DESCRIPTION
**Pull request by jnewbery**
Adds an `updatepeer` RPC to update peer config and carry out actions on the peer by NodeId. At the moment, the actions are just changing whether the peer is whitelisted and whether the peer is a manual_connection (previously called an addnode). Future possible actions would be changing banscore, banning, and so on.

This is designed to be called using named arguments, but due to the RPC infrastructure can also be called with positional arguments (although doing so would be very fiddly).

I've set the category to hidden for now. We may want to make the whitelisting behaviour more granular in future and I don't want to commit us to a public API that we can't then change.

### CODE REVIEW

**Comment by theuni** [comment]
I'm not sure I like the idea of adding a command that changes the implied state of a peer via its attributes, rather than performing actions explicitly. The whitelist side-effects are especially hazy, see the brief discussion here: https://github.com/bitcoin/bitcoin/pull/10051#issuecomment-290622674.

**Comment by jonasschnelli** [comment]
@jnewbery:
Can you elaborate the use-case for `disconnect`? IMO `setban` provides a similar interface, with disconnecting & banning for a specific timespan (`1h`, etc.) because, a pure `disconnect` does not prevent the peer from a direct re-connect.

**Comment by jonasschnelli** [comment]
Ah... an there is already `disconnectnode`.

**Comment by jnewbery** [comment]
@jonasschnelli see #2729 and #6271 for history of the `disconnectnode` RPC. It's also useful in testing to be able to control the topology of the test nodes.

@theuni - I agree that whitelisting is a mess and should be broken out into bits for controlling individual behaviours. Hence my original comment: "I have a feeling we may want to make the whitelisting behaviour more granular and I don't want to commit us to a public API that we can't then change." I was thinking of your comment in #10051 but couldn't find the reference.

> the idea of adding a command that changes the implied state of a peer via its attributes, rather than performing actions explicitly

I don't understand what this means in the context of whitelisting. The idea is to update the peer's whitelist behaviour. What would performing actions explicitly entail?

**Comment by jnewbery** [comment]
Closing in favour of #10143 for the immediate need (disconnect node by id), but I think this could still be a useful RPC in the future.

**Comment by jnewbery** [comment]
Reopening with just the ability to update:

- `fWhitelisted`
- `m_manual_connection`

The net_processing functionality for the v0.15.0.2 PRs is disabled for manual connections, so this PR could be helpful for testing those changes.

@sdaftuar

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -121,73 +182,79 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
+            throw JSONRPCError(RPC_MISC_ERROR, "Failed to update node whitelisting");
         }
-        obj.push_back(Pair("bytessent_per_msg", sendPerMsgCmd));
+    }
 
-        UniValue recvPerMsgCmd(UniValue::VOBJ);
-        for (const mapMsgCmdSize::value_type &i : stats.mapRecvBytesPerMsgCmd) {
-            if (i.second > 0)
-                recvPerMsgCmd.push_back(Pair(i.first, i.second));
+    if (request.params.size() >= 3) {
```
Comment (promag) [inline src/rpc/net.cpp]: Is null instead.

File: src/rpc/net.cpp [hunk 2]
```diff
@@ -121,73 +182,79 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
+            throw JSONRPCError(RPC_MISC_ERROR, "Failed to update node manual");
         }
-        obj.push_back(Pair("bytesrecv_per_msg", recvPerMsgCmd));
+    }
 
-        ret.push_back(obj);
+    UniValue entry(UniValue::VOBJ);
+    CNodeStats stats;
+
+    if (g_connman->GetNodeStats(nodeid, stats)) {
```
Comment (promag) [inline src/rpc/net.cpp]: If not then raise invalid parameter?

Comment (promag) [reply src/rpc/net.cpp]: Missing test for invalid node.

File: src/rpc/net.cpp [hunk 3]
```diff
@@ -121,73 +182,79 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
     if(!g_connman)
         throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
 
-    std::vector<CNodeStats> vstats;
-    g_connman->GetNodeStats(vstats);
 
     UniValue ret(UniValue::VARR);
 
-    for (const CNodeStats& stats : vstats) {
+    if (request.params.size() >= 1 && request.params[0].isNum()) {
```
Comment (promag) [inline src/rpc/net.cpp]: Not null instead.

**Review by promag — COMMENTED** [review]
Concept ACK. Will test.

**Comment by NicolasDorier** [comment]
whitelisting via rpc?? BIG BIG concept ACK. It will make configuration of services depending on Bitcoin Core RPC and P2P so much easier.

File: src/rpc/net.cpp [hunk 4]
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
-            obj.push_back(Pair("synced_blocks", statestats.nCommonHeight));
-            UniValue heights(UniValue::VARR);
-            for (int height : statestats.vHeightInFlight) {
-                heights.push_back(height);
-            }
-            obj.push_back(Pair("inflight", heights));
+        CNodeStats stats;
+
+        int node_id = request.params[0].get_int();
+        if (g_connman->GetNodeStats(node_id, stats)) {
```
Comment (promag) [inline src/rpc/net.cpp]:
```cpp
if (!...) {
    throw ...;
}
```

File: src/rpc/net.cpp [hunk 5]
```diff
@@ -65,12 +65,73 @@ UniValue ping(const JSONRPCRequest& request)
...
+        if (i.second > 0)
+            recvPerMsgCmd.push_back(Pair(i.first, i.second));
+    }
+    obj.push_back(Pair("bytesrecv_per_msg", recvPerMsgCmd));
+}
+
 UniValue getpeerinfo(const JSONRPCRequest& request)
 {
-    if (request.fHelp || request.params.size() != 0)
+    if (request.fHelp || request.params.size() >= 2)
```
Comment (promag) [inline src/rpc/net.cpp]: Nit, `> 1`.

File: src/rpc/net.cpp [hunk 6]
```diff
@@ -65,12 +65,73 @@ UniValue ping(const JSONRPCRequest& request)
...
+        for (int height : statestats.vHeightInFlight) {
+            heights.push_back(height);
+        }
+        obj.push_back(Pair("inflight", heights));
+    }
+    obj.push_back(Pair("whitelisted", stats.fWhitelisted));
+
+    UniValue sendPerMsgCmd(UniValue::VOBJ);
+    for (const mapMsgCmdSize::value_type &i : stats.mapSendBytesPerMsgCmd) {
+        if (i.second > 0)
```
Comment (promag) [inline src/rpc/net.cpp]: Nit, same line or `{ }`.

File: src/net.cpp [hunk 7]
```diff
@@ -2521,6 +2521,18 @@ size_t CConnman::GetNodeCount(NumConnections flags)
     return nNum;
 }
 
+bool CConnman::GetNodeStats(NodeId nodeid, CNodeStats& stats)
+{
+    LOCK(cs_vNodes);
+    for(CNode* pnode : vNodes) {
```
Comment (promag) [inline src/net.cpp]: Nit, missing spaces after `for`.

File: src/rpc/net.cpp [hunk 8]
```diff
@@ -65,12 +65,73 @@ UniValue ping(const JSONRPCRequest& request)
...
+            sendPerMsgCmd.push_back(Pair(i.first, i.second));
+    }
+    obj.push_back(Pair("bytessent_per_msg", sendPerMsgCmd));
+
+    UniValue recvPerMsgCmd(UniValue::VOBJ);
+    for (const mapMsgCmdSize::value_type &i : stats.mapRecvBytesPerMsgCmd) {
+        if (i.second > 0)
```
Comment (promag) [inline src/rpc/net.cpp]: Same as above.

File: src/rpc/net.cpp [hunk 9]
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
-        UniValue sendPerMsgCmd(UniValue::VOBJ);
-        for (const mapMsgCmdSize::value_type &i : stats.mapSendBytesPerMsgCmd) {
-            if (i.second > 0)
-                sendPerMsgCmd.push_back(Pair(i.first, i.second));
+    return ret;
+}
+
+UniValue updatepeer(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() > 3 || request.params.size() == 0)
```
Comment (promag) [inline src/rpc/net.cpp]: Nit, the most common expression is
```cpp
if (request.fHelp || request.params.size() < 1 || request.params.size() > 3)
```

File: src/rpc/net.cpp [hunk 10]
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
+            "\nArguments:\n"
+            "1. \"node\"              (string, required) The node id (see getpeerinfo for a list of nodes with their ids)\n"
+            "2. \"whitelisted\"       (bool, optional) whether the node is whitelisted\n"
+            "3. \"manual_connection\" (bool, optional) whether the node was manually added\n"
+            "\nExamples:\n"
+            + HelpExampleCli("updatepeer", "0 true")
+            + HelpExampleRpc("updatepeer", "0, true")
+        );
+
+    if(!g_connman) {
```
Comment (promag) [inline src/rpc/net.cpp]: Nit, space after `if`.

File: src/rpc/net.cpp [hunk 11]
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
+        throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
+    }
+
+    LOCK(cs_main);
+
+    RPCTypeCheckArgument(request.params[0], UniValue::VNUM);
+    int nodeid = request.params[0].get_int();
```
Comment (promag) [inline src/rpc/net.cpp]: Same validation of `getpeerinfo`.

File: src/rpc/net.cpp (hunk 5 above)
Comment (promag) [inline src/rpc/net.cpp]: Needs release notes.

File: src/rpc/net.cpp [hunk 12]
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
-            obj.push_back(Pair("synced_headers", statestats.nSyncHeight));
-            obj.push_back(Pair("synced_blocks", statestats.nCommonHeight));
-            UniValue heights(UniValue::VARR);
-            for (int height : statestats.vHeightInFlight) {
-                heights.push_back(height);
-            }
-            obj.push_back(Pair("inflight", heights));
+        CNodeStats stats;
+
+        int node_id = request.params[0].get_int();
```
Comment (promag) [inline src/rpc/net.cpp]: Actually I think this is more correct:
 - raise `RPC_INVALID_PARAMETER` if `node_id < 0`;
 - raise `RPC_INVALID_ADDRESS_OR_KEY` if `!GetNodeStats(...)`.

Note: if you do this then add a test for 1st case.

File: test/functional/net.py [hunk 13]
```diff
@@ -94,5 +95,31 @@ def _test_getpeerinfo(self):
...
+        assert peer_info['addnode']
+
+        # Update properties and verify that they've changed
+        self.nodes[0].updatepeer(peer_info['id'], True, False)
+        peer_info = self.nodes[0].getpeerinfo()[0]
+        assert peer_info['whitelisted']
+        assert not peer_info['addnode']
+
+        # Check that updatepeer fails for an invalid peer_id
+        assert_raises_rpc_error(-8, "not found", self.nodes[0].updatepeer, 5000)
```
Comment (promag) [inline test/functional/net.py]: Missing failure tests for getpeerinfo (maybe unrelated to this PR).

**Review by promag — COMMENTED** [review]
Nit, rename PR to *Add updatepeer RPC*. Squash `[RPC] Allow manual_connection to be updated using updatepeer` into `[RPC] Add updatenode RPC`.

File: src/rpc/net.cpp (hunk 12 above)
Comment (jnewbery) [reply src/rpc/net.cpp]: I don't think it's too important, but feel free to open a follow-up PR if you think it's required.

**Comment by jnewbery** [comment]
@promag nits addressed in latest commit.

File: src/rpc/net.cpp [hunk 14]
```diff
@@ -121,73 +184,82 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
+            "\nExamples:\n"
+            + HelpExampleCli("updatepeer", "0 true")
+            + HelpExampleRpc("updatepeer", "0, true")
+        );
+
+    if (!g_connman) {
+        throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
+    }
+
+    LOCK(cs_main);
```
Comment (promag) [inline src/rpc/net.cpp]: Remove lock or am I missing something?

**Comment by promag** [comment]
utACK after squash.

Nit, how about `updatepeer id setting value (setting value ...)`:
```
updatepeer 1 whitelisted true
updatepeer 1 manual_connection false
```
It works well without named arguments with the cli, and scales well for other settings.

Not, I always prefer named arguments, specially in these calls. But we do support index arguments and this approach doesn't help for future settings.

**Review by promag — COMMENTED** [review]
utACK after squash.

Nit, how about `updatepeer id setting value (setting value ...)`:
```
updatepeer 1 whitelisted true
updatepeer 1 manual_connection false
```
It works well without named arguments with the cli, and scales well for other settings.

Note, I always prefer named arguments, specially in these calls. But we do support index arguments and this approach doesn't help for future settings.

**Comment by NicolasDorier** [comment]
I agree with @promag `updatepeer id setting value (setting value ...)` would be easier to extend later.

**Comment by jnewbery** [comment]
> Nit, how about updatepeer id setting value (setting value ...)

This is a different scheme from all of the existing RPCs. I expect that there would need to be changes to the rpc framework to make this work.

It doesn't make sense to me have a completely different scheme for just this RPC method, since we already have named arguments.

If you disagree, perhaps you could implement the scheme you're talking about in a new branch?

**Comment by NicolasDorier** [comment]
@jnewbery how so? This is just 3 differents parameters of 3 strings.

The way you are doing now, if there is like 60 settings for one peer, we will have 60 parameters to this function. This does not seem very maintainable. It would be better in such case to have `updatewhitelist id true`.

**Comment by promag** [comment]
> This is designed to be called using named arguments, but due to the RPC infrastructure can also be called with positional arguments (although doing so would be very fiddly).

@jnewbery you too agree that is by design bad for positional arguments (considering the possible settings can be extended).

Another approach is to use something like `sendmany`: `updatepeer {\"setting\":value, ...}`.

FWIW I already gave my utACK 😄

**Comment by jnewbery** [comment]
@NicolasDorier 

> how so? This is just 3 differents parameters of 3 strings

Right, but all RPCs currently support positional arguments, where the position of the argument determines its meaning. I don't necessarily agree that it's always appropriate, but having a single RPC that doesn't adhere to that scheme is confusing and inconsistent.

Using named arguments have been around for at least two releases now, so there's no need to use positional arguments at all.

Rebased

**Comment by NicolasDorier** [comment]
yeah we talked about it, I understand now why it makes sense named args. Concept ACK.

**Comment by jnewbery** [comment]
@NicolasDorier - you've given this a *BIG BIG concept ACK*. Mind reviewing? :slightly_smiling_face:

**Comment by NicolasDorier** [comment]
Sure will review and test today.

File: src/net.cpp [hunk 15]
```diff
@@ -2571,6 +2583,30 @@ bool CConnman::DisconnectNode(NodeId id)
...
+            return true;
+        }
+    }
+    return false;
+}
+
+bool CConnman::SetManualConnection(NodeId id, bool manual)
+{
+    LOCK(cs_vNodes);
+    for(CNode* pnode : vNodes) {
```
Comment (NicolasDorier) [inline src/net.cpp]: The loop will get repeated all over the place for each field. (Already two times)

I think a better approach would be to lock at `updatepeer` level as well as fetch once for all the right CNode. And change fields directly from there.

File: src/rpc/net.cpp [hunk 16]
```diff
@@ -121,73 +184,82 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
+
+    CNodeStats stats;
+    if (!g_connman->GetNodeStats(node_id, stats)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Node id %u not found", node_id));
+    }
+
+    if (!request.params[1].isNull()) {
+        RPCTypeCheckArgument(request.params[1], UniValue::VBOOL);
+        if (!g_connman->SetWhitelisted(node_id, request.params[1].isTrue())) {
+            throw JSONRPCError(RPC_MISC_ERROR, "Failed to update peer whitelisting");
```
Comment (NicolasDorier) [inline src/rpc/net.cpp]: See my above comment, by fetching once the node at the start of this method, you can remove all those `if` as you already know if the CNode with such id already exist.

File: src/net.cpp [hunk 17]
```diff
@@ -2539,6 +2539,18 @@ size_t CConnman::GetNodeCount(NumConnections flags)
     return nNum;
 }
 
+bool CConnman::GetNodeStats(NodeId nodeid, CNodeStats& stats)
```
Comment (NicolasDorier) [inline src/net.cpp]: This method would also be removed if fetching CNode was done at the updatepeer level.

**Comment by NicolasDorier** [comment]
@jnewbery do you need more help on this PR? I am really interested into seeing it merged.

File: src/net.cpp [hunk 18]
```diff
@@ -2553,6 +2565,30 @@ bool CConnman::DisconnectNode(NodeId id)
     return false;
 }
 
+bool CConnman::SetWhitelisted(NodeId id, bool fWhitelisted)
+{
+    LOCK(cs_vNodes);
+    for(CNode* pnode : vNodes) {
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/net.cpp]: fixed

File: src/rpc/net.cpp (hunk 14 above)
Comment (jnewbery) [reply src/rpc/net.cpp]: you're right - I don't think this is required

**Comment by jnewbery** [comment]
@NicolasDorier: Sorry for dropping this - I've rebased on master and squashed all nits.

I'm not sure about your suggestion for locking in updatepeer and fetching the `CNode`. `cs_vNodes` isn't currently locked anywhere outsdie `CConman`, which I think is a good property to maintain.

**Comment by NicolasDorier** [comment]
@jnewbery in this case, I would suggest to have a method on CConman `UpdateSetting(str,value)` because there is high ratio of monkey copy pasta code everytimes we will need to add one property.

**Comment by TheBlueMatt** [comment]
I'm still with @theuni on this one, not super happy with the idea of changing properties about peers that net_processing/net both consider "constant". Would prefer we add some (undocumented?) options to addnode (or a new RPC) which lets you control the flags of a peer as you create a connection.

**Comment by jnewbery** [comment]
@theuni's comment was:

> the idea of adding a command that changes the implied state of a peer via its attributes, rather than performing actions explicitly

I don't understand what's meant by 'performing actions explicitly', but I'm happy to modify this PR if there are concrete suggestions.

> not super happy with the idea of changing properties about peers that net_processing/net both consider "constant"

Can you articulate what makes you not happy? Being able to change the properties of a connection without having to delete/recreate that object is generally very useful, since cycling a connection has many side-effects. As far as I can see, the only stateful impact of starting with `fWhitelisted` rather than updating it later is this call to `AddLocal()` in `BindListenPort()`:

```
    if (addrBind.IsRoutable() && fDiscover && !fWhitelisted)
        AddLocal(addrBind, LOCAL_BIND);
```

I can't see any other reason to be concerned about updating `fWhitelisted` dynamically.

**Comment by TheBlueMatt** [comment]
I might be slightly more OK with changing fWhitelisted on a peer after its up, as that may be useful "in the real world" outside of tests, but I'm super not a fan of changing manual_connection. One thing @theuni and I have talked about is being more aggressive about CNode representing a "connection handle" where its potentially allowed to have a few constant members which are general information about the connection, eg whether it was automatically or manually added, the remote address, etc.
One thing you might imagine is inserting a peer into a set/map (eg compact blocks HB mode peers set, the (implied) set of nPreferredDownload peers (which we currently track with fPreferredDownload, but we could actually take it as implication based on const values in CNode)) or otherwise performing (between-message-)stateful behavior on a peer based on such (constant) information about a connection. In each of these cases, having such constant information change out from under you could potentially introduce races/bugs.

> cycling a connection has many side-effects.
I'm not sure what side-effects there are that couldn't be reproduced on a new connection after cycling the connection. It ends up being an unrealistic test if you dont do that anyway. Maybe I'm just missing the motivation of cases that wouldnt be better tested by cycling a connection?

**Comment by jnewbery** [comment]
> One thing you might imagine...

You're describing things that _may_ happen in future changes in abstract terms, particularly when you talk about between-message-stateful behaviour.

`m_manual_connection` is certainly a better name than `fAddNode`, but I still don't think it really captures what the property means. Really it's something like 'preferred_peer' - it's a peer that we don't want to disconnect or punish for bad behaviour. How it was connected isn't really relevant. We may want to prefer a peer that has connected inbound to us, and we may want to connect outbound to a peer without preferring it. In that context, I think it makes sense to want to set this property manually after connection, without having to disconnect and go through the version-verack-getheaders-headers handshaking.

This is undoubtedly useful for testing, but I could also imagine it being useful, for example in a business that has more than one bitcoin node and wants to be able to manage their network topology.

Thinking forward a bit more, I think it'd also be beneficial to break out the whitelist behaviour to be more granular, and again be able to update those properties dynamically.

**Review by ryanofsky — COMMENTED** [review]
Haven't been following discussion in much detail but if the problem is `fAddNode` having both const and non-const interpretations, maybe it should just be split into two variables: `m_connection_type` that would be const and record origin of connection and `m_preferred_peer` that would be mutable.

File: src/net.cpp [hunk 19]
```diff
@@ -2577,6 +2589,30 @@ bool CConnman::DisconnectNode(NodeId id)
...
+    LOCK(cs_vNodes);
+    for (CNode* pnode : vNodes) {
+        if (id == pnode->id) {
+            pnode->m_manual_connection = manual;
+            return true;
+        }
+    }
+    return false;
+}
+
```
Comment (kallewoof) [inline src/net.cpp]: Could you maybe make an (optionally inline) method `find_node_by_id` that the above methods use? Looks a bit copy-pastey.

File: src/rpc/net.cpp [hunk 20]
```diff
@@ -65,12 +65,80 @@ UniValue ping(const JSONRPCRequest& request)
...
 UniValue getpeerinfo(const JSONRPCRequest& request)
 {
-    if (request.fHelp || request.params.size() != 0)
+    if (request.fHelp || request.params.size() > 1)
         throw std::runtime_error(
             "getpeerinfo\n"
-            "\nReturns data about each connected network node as a json array of objects.\n"
+            "\nReturns data about connected network nodes as a json array of objects.\n"
+            "\nArguments:\n"
+            "1. \"id\"         (numeric, optional) Only return information about the peer with the specified id.\n"
```
Comment (kallewoof) [inline src/rpc/net.cpp]: Nit: No period at end (seems to be the norm). Also add this to the first line (`"getpeerinfo \"id\"\n"`)

File: src/rpc/net.cpp [hunk 21]
```diff
@@ -121,73 +189,80 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
+    }
+
+    return ret;
+}
+
+UniValue updatepeer(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 3)
+        throw std::runtime_error(
+            "updatepeer\n"
```
Comment (kallewoof) [inline src/rpc/net.cpp]: `"updatepeer \"id\" ( \"whitelisted\" \"manual_connection\" )\n"`

File: src/rpc/net.cpp [hunk 22]
```diff
@@ -121,73 +189,80 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
...
+    if (!g_connman) {
+        throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
+    }
+
+    RPCTypeCheckArgument(request.params[0], UniValue::VNUM);
+    int node_id = request.params[0].get_int();
+
+    CNodeStats stats;
+    if (!g_connman->GetNodeStats(node_id, stats)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Node id %u not found", node_id));
```
Comment (kallewoof) [inline src/rpc/net.cpp]: Nit: `node_id` is an `int`, so `%d` not `%u`.

**Review by kallewoof — COMMENTED** [review]
utACK

Needs rebase.

File: src/net.cpp (hunk 19 above)
Comment (jnewbery) [reply src/net.cpp]: refactored out

File: src/rpc/net.cpp (hunk 20 above)
Comment (jnewbery) [reply src/rpc/net.cpp]: fixed

File: src/rpc/net.cpp (hunk 22 above)
Comment (jnewbery) [reply src/rpc/net.cpp]: fixed

**Comment by jnewbery** [comment]
Fixed @kallewoof review comments.

**Comment by jnewbery** [comment]
There doesn't seem to much appetite for this. Closing.

**Comment by NicolasDorier** [comment]
I liked it :(

**Comment by MarcoFalke** [comment]
Marked "up for grabs". (Conversation is locked, so a new pull will need to be opened either way).

**Pull request closed without merging**